### ISSUE TITLE
One-assertion-per-test rule

### ISSUE THREAD

**Issue by chillenzer** [edbennett/python-testing-ci#11]
As a good rule of thumb, I learnt that there should be only a single assertion per test. This will ensure that whenever a test fails

* you know precisely which result was not as expected
* the other tests will still run (with usual settings in standard test frameworks) potentially yielding additional information
* you can rely on the fact that there cannot be any unintended state-changing behaviour before that single assertion.

As with all these rules, this is not precisely to be taken literally. Sometimes a collective check, e.g. testing all neighbours of a cell, is spiritually to be seen as a single assertion. In such cases -- particularly if this is a repeating pattern for a large number of tests because, e.g., you want to test the same kind of assertions for lots of different inputs -- it is advised IMHO to wrap this set of assertions into a function itself to clarify the intent. Also, from my perspective one should make sure that the set of assertions is really conceptually related.

I bring this up for two reasons: First, shall we discuss this concept in the course? Second, your `grid` example violates this rule. As I hinted before, I believe that this might be okay for the bulk test as it is conceptually very consistent but I would disagree in saying that the edge tests can be seen that way. I would personally write a set of tests like this probably:
```python
def test_left_edge_has_three_neighbours():
    [...]
    assert cell.neighbours() == 3

def test_left_edge_has_no_left_neighbour():
    [...]
    assert cell.left() is None
    
def test_left_edge_has_correct_bulk_behaviour():
    [...]
    assert <the three others>
```
and so on.

**Comment by edbennett** [comment]
Good point. It would probably be worth discussing this; it definitely helps when trying to pin down the exact cause of bugs. You would end up with quite a large number of tests, but we don't have to type them out so that's not a disaster. I guess the question is whether it would introduce too many concepts at once—would it need a separate episode, or section of episode?

(Small point: this material isn't "mine"; was developed by the Bristol RSE team, I just massaged it into the Carpentries template.)
